File: go/src/v.io/x/browser/sample/mocks/poolheater.go
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package mocks

import (
	"time"

	"v.io/v23/context"
	"v.io/v23/rpc"
)

const (
	// Pool heater status constants
	poolHeaterActive = "active"
	poolHeaterIdle   = "idle"

	poolHeaterDefaultTemperature = 60
)

// PoolHeater allows clients to control when the pool is being heated.
type poolHeater struct {
	status          string
	currTemperature uint64
}

// Status retrieves the PoolHeater's status (i.e., active, idle) and temperature.
func (p *poolHeater) Status(*context.T, rpc.ServerCall) (status string, temperature uint64, err error) {
	return p.status, p.currTemperature, nil
}

// Start informs the PoolHeater to heat the pool to the given temperature until the duration expires.
func (p *poolHeater) Start(_ *context.T, _ rpc.ServerCall, temperature uint64, duration uint64) error {
	// Begin heating.
	p.status = poolHeaterActive
	p.currTemperature = temperature

	// After duration, stop heating.
	time.AfterFunc(
		time.Duration(duration)*time.Second,
		func() {
			p.status = poolHeaterIdle
			p.currTemperature = poolHeaterDefaultTemperature
		},
	)
	return nil
}

// Stop informs the PoolHeater to cease heating the pool.
func (p *poolHeater) Stop(*context.T, rpc.ServerCall) error {
	p.status = poolHeaterIdle
	p.currTemperature = poolHeaterDefaultTemperature
	return nil
}

// NewPoolHeater creates a new poolHeater stub.
func NewPoolHeater() *poolHeater {
	return &poolHeater{
		status:          poolHeaterIdle,
		currTemperature: poolHeaterDefaultTemperature,
	}
}

File: src/services/sample-world/poolheater.js
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

var vdl = require('./ifc');

module.exports = PoolHeater;

var ACTIVE = 'active';
var IDLE = 'idle';

var DEFAULT_TEMPERATURE = 60;

// PoolHeater allows clients to control when the pool is being heated.
function PoolHeater() {
  this.state = IDLE;
  this.currTemperature = DEFAULT_TEMPERATURE;
  this.handle = null;
}

PoolHeater.prototype = new vdl.PoolHeater();

// Status retrieves the PoolHeater's status (i.e., active, idle) and temperature
PoolHeater.prototype.status = function(context, serverCall) {
  return [this.state, this.currTemperature];
};

// Start informs the PoolHeater to heat the pool to the given temperature until
// the duration expires.
PoolHeater.prototype.start =
  function(context, serverCall, temperature, duration) {

  clearTimeout(this.handle); // Clear timeout handle, if any.

  // Begin heating.
  this.status = ACTIVE;
  this.currTemperature = temperature;

  // After duration, stop heating.
  var self = this;
  this.handle = setTimeout(function() {
    self.state = IDLE;
    self.currTemperature = DEFAULT_TEMPERATURE;
  }, duration * 1000);
};

// Stop informs the PoolHeater to cease heating the pool.
PoolHeater.prototype.stop = function(context, serverCall) {
  clearTimeout(this.handle); // Clear timeout handle, if any.

  this.state = IDLE;
  this.currTemperature = DEFAULT_TEMPERATURE;
};